### ISSUE TITLE
Sort sentences by topic: using LDA/PCA to remix novels/twitter.

### ISSUE THREAD

**Issue by Cortexelus** [dariusk/NaNoGenMo#77]
Topic modeling is easy with Python and Gensim: http://radimrehurek.com/gensim/index.html

For example, I could take a novel and convert its sentences to bag-of-words vectors. 
[See this tutorial http://radimrehurek.com/gensim/tut1.html]

Then I could perform PCA on all the vectors, extract the principle component, and sort by it. This would arrange sentences along a topic path. Sentences talking about the same stuff clump together, and as you read it gradually moves onto talking about different stuff. 

I think it'd be interesting for the twitter novels to sort their tweets this way. 

Or I could group sentences together by topic, and name each chapter by the most prominent phrases/words in each topic. 

Or I could leverage WordNet or ConceptNet to do a more clever topic-modeler. (This gives me knowledge, for example, that "dog" and "cat" are both "animals", and so sentences about dogs and cats may be grouped together in a topic). One approach: for each word in each bag-of-words vector, add its hypernym or/and its hypernym's hypernym to the bag.

Topic modeling with LDA: http://en.wikipedia.org/wiki/Latent_Dirichlet_allocation
Dimensionality reduction with PCA: http://en.wikipedia.org/wiki/Principal_component_analysis

``` python
# PCA is easy in python with sklearn:
from sklearn import decomposition
data = [[1,2,4,5],[7,2,1,2],[5,7,8,2]]
pca = decomposition.PCA(n_components=1) # reduce to one dimension
pca.fit(data)
data_reduced_to_one_dimension = pca.transform(data)
```

**Comment by enkiv2** [comment]
This is a pretty cool idea, although if I was going to bring in something
as heavy duty as ConceptNet I'd probably generate the text directly from
that knowledgebase rather than using it as a sub-processor.

(Similarly, somebody was using WordNet to shift words from more specific to
more general meanings.)

**Comment by Cortexelus** [comment]
Right. One way to use WordNet in topic modeling is to take each bag-of-words vector and swap each word in with its hypernym, or its hypernym's hypernym. Actually, instead of swapping, you could just add the hypernym into the bag. That might work well to improve the topic modeler.

**Comment by Cortexelus** [comment]
Turns out PCA is awful way to sort by topic, as it finds a linear solution, whereas related sentences in 2500-dimensional topic space form an especially wiggly manifold. Instead I am running Traveling Salesman Problem approximation using simulated annealing. Turns out this is especially slow. I anticipate this needing to run for several days before it spits out an approximation. Total distance has only shrunk to 80% of random since I started running it last night. Ideally it should shrink to around 2%.

**Comment by ikarth** [comment]
I'll be interested in seeing what your results are, since one idea I had to improve my generator was to find a way to group the sentences together and I haven't worked out a good way to do that yet.

**Comment by Cortexelus** [comment]
Traveling Salesman on 2500-dimensional space ran for a day, improved its distance by 2/3rds, and got stuck in a local minimum. I must try a compromise. 
- Sorting by the first principal component (PCA with one dimension) does organize some topics globally, but doesn't preserve local relatedness.
- Traveling Salesman should maximize local relatedness, but it takes too long with 5200 points in 2500-dimensional word-vectors, and heuristics easily get trapped in local minima. 

What I need to do is reduce dimensionality (k = 50-500), then do Traveling Salesman. 
- I could cluster sentences by topic (Kmeans, or LDA, or LDA+Kmeans), do Traveling Salesman within each topic, and then do Traveling Salesman on the topic clusters themselves. 
- OR I could do PCA with k components, then do Traveling Salesman on each sentence's principle component vector.
- OR I could do LDA with k topics, then do Traveling Salesman on each sentence's topic vector.

**Comment by MichaelPaulukonis** [comment]
@Cortexelus - did you get anywhere with this?
